enterprise_access/apps/subsidy_access_policy/models.py

```diff
@@ -1509,3 +1511,148 @@ class Meta:
+    (and related enrollment/fulfillment) of a particular learner,
+    covered by a particular subsidy access policy, into some specific course run.
+    This needs exists for reasons related to upstream business constraints,
+    notably in cases where a course is included in a policy's catalog,
+    but the desired course run is not discoverable due to the
+    current state of its metadata. This model supports executing such a redemption.
+
+    .. no_pii: This model has no PII
+    """
+    uuid = models.UUIDField(
+        primary_key=True,
+        default=uuid4,
+        editable=False,
+        unique=True,
+        help_text='The uuid that uniquely identifies this policy record.',
+    )
+    subsidy_access_policy = models.ForeignKey(
+        SubsidyAccessPolicy,
+        related_name="forced_redemptions",
+        on_delete=models.SET_NULL,
+        null=True,
+        help_text="The SubsidyAccessPolicy that this forced redemption relates to.",
+    )
+    lms_user_id = models.IntegerField(
+        null=False,
+        blank=False,
+        db_index=True,
+        help_text=(
+            "The id of the Open edX LMS user record that identifies the learner.",
+        ),
+    )
+    course_run_key = models.CharField(
+        max_length=255,
+        blank=False,
+        null=False,
+        db_index=True,
+        help_text=(
+            "The course run key to enroll the learner into.",
+        ),
+    )
+    content_price_cents = models.BigIntegerField(
+        null=False,
+        blank=False,
+        help_text="Cost of the content in USD Cents, should be >= 0.",
+    )
+    wait_to_redeem = models.BooleanField(
+        default=False,
+        help_text="If selected, will not force redemption when the record is saved via Django admin.",
+    )
+    redeemed_at = models.DateTimeField(
+        null=True,
+        blank=True,
+        help_text="The time the forced redemption succeeded.",
+    )
+    errored_at = models.DateTimeField(
+        null=True,
+        blank=True,
+        help_text="The time the forced redemption failed.",
+    )
+    traceback = models.TextField(
+        blank=True,
+        null=True,
+        editable=False,
+        help_text="Any traceback we recorded when an error was encountered.",
+    )
+    transaction_uuid = models.UUIDField(
+        null=True,
+        db_index=True,
+        help_text=(
+            "The transaction uuid caused by successful redemption.",
+        ),
+    )
+    history = HistoricalRecords()
+
+    def __str__(self):
+        return (
+            f'policy_uuid={self.subsidy_access_policy.uuid}, transaction_uuid={self.transaction_uuid}, '
+            f'lms_user_id={self.lms_user_id}, course_run_key={self.course_run_key}'
+        )
+
+    def create_assignment(self):
+        """
+        For assignment-based policies, an allocated ``LearnerContentAssignment`` must exist
+        before redemption can occur.
+        """
```

Comment:
Not entirely sure what the plan is for assignments, but I assume this logic will change later to support assigning course run keys?

Reply:
Oh, I should probably read the ADR later...